src/main/java/com/redhat/rhjmc/containerjfr/platform/ServiceRef.java

```diff
@@ -41,36 +41,45 @@
  */
 package com.redhat.rhjmc.containerjfr.platform;
 
+import java.net.MalformedURLException;
+import java.util.Optional;
+
+import javax.management.remote.JMXServiceURL;
+
 import org.apache.commons.lang.builder.EqualsBuilder;
 import org.apache.commons.lang.builder.HashCodeBuilder;
 import org.apache.commons.lang3.builder.ToStringBuilder;
+import com.google.gson.annotations.SerializedName;
 
 public class ServiceRef {
 
-    private final String connectUrl;
-    private final String alias;
-    private final int port;
+    @SerializedName("connectUrl")
+    private final JMXServiceURL JMXServiceURL;
 
-    public ServiceRef(String connectUrl, int port) {
-        this(connectUrl, connectUrl, port);
-    }
+    private final String alias; // nullable
 
-    public ServiceRef(String connectUrl, String alias, int port) {
-        this.connectUrl = connectUrl;
+    public ServiceRef(JMXServiceURL jmxServiceUrl, String alias) throws MalformedURLException {
+        this.JMXServiceURL = jmxServiceUrl;
         this.alias = alias;
-        this.port = port;
     }
 
-    public String getConnectUrl() {
-        return connectUrl;
+    public ServiceRef(JMXServiceURL jmxServiceUrl) throws MalformedURLException {
+        this(jmxServiceUrl, null);
+    }
+
+    public ServiceRef(String host, int port, String alias) throws MalformedURLException {
+        this(
+                new JMXServiceURL(
+                        "rmi", "", 0, String.format("/jndi/rmi://%s:%s/jmxrmi", host, port)),
```

Comment:
Ah, tiny nitpick here - your format specifier has two `%s` tokens, but you pass in a string (`%s`) and an int (`%d`). This will still work because casting the port to a string still comes out the same in the formatted string, but I think it's still worth correcting.

Reply:
Ah true, Ill fix it!